src/transaction/log_replication_atomic.cpp

```diff
@@ -0,0 +1,173 @@
+	  case LOG_MVCC_DIFF_UNDOREDO_DATA:
+	    read_and_redo_record_with_atomic_consideration<log_rec_mvcc_undoredo> (thread_entry, header.type, header.back_lsa,
+		m_redo_lsa, header.trid);
+	    break;
+	  case LOG_RUN_POSTPONE:
+	    read_and_redo_record_with_atomic_consideration<log_rec_run_postpone> (thread_entry, header.type, header.back_lsa,
+		m_redo_lsa, header.trid);
+	    break;
+	  case LOG_COMPENSATE:
+	    read_and_redo_record_with_atomic_consideration<log_rec_compensate> (thread_entry, header.type, header.back_lsa,
+		m_redo_lsa, header.trid);
+	    break;
+	  case LOG_DBEXTERN_REDO_DATA:
+	  {
+	    const log_rec_dbout_redo dbout_redo =
+		    m_redo_context.m_reader.reinterpret_copy_and_add_align<log_rec_dbout_redo> ();
+	    log_rcv rcv;
+	    rcv.length = dbout_redo.length;
+
+	    log_rv_redo_record (&thread_entry, m_redo_context.m_reader, RV_fun[dbout_redo.rcvindex].redofun, &rcv,
+				&m_redo_lsa, 0, nullptr, m_redo_context.m_redo_zip);
+	    break;
+	  }
+	  case LOG_COMMIT:
+	    calculate_replication_delay_or_dispatch_async<log_rec_donetime> (
+		    thread_entry, m_redo_lsa);
+	    break;
+	  case LOG_ABORT:
+	    calculate_replication_delay_or_dispatch_async<log_rec_donetime> (
+		    thread_entry, m_redo_lsa);
+	    break;
+	  case LOG_DUMMY_HA_SERVER_STATE:
+	    calculate_replication_delay_or_dispatch_async<log_rec_ha_server_state> (
+		    thread_entry, m_redo_lsa);
+	    break;
+	  case LOG_TRANTABLE_SNAPSHOT:
+	    // save the LSA of the last transaction table snapshot that can be found in the log
+	    // only needed on the passive transaction server
+	    m_most_recent_trantable_snapshot_lsa.store (m_redo_lsa);
+	    break;
+	  case LOG_START_ATOMIC_REPL:
+	  case LOG_SYSOP_ATOMIC_START:
+	    if (m_atomic_helper.is_part_of_atomic_replication (header.trid))
+	      {
+		// nested atomic replication
+		assert (false);
+	      }
+	    m_atomic_helper.start_new_atomic_replication_sequence (header.trid);
+	    break;
+	  case LOG_END_ATOMIC_REPL:
+	  case LOG_SYSOP_END:
+	    if (!m_atomic_helper.is_part_of_atomic_replication (header.trid))
+	      {
+		//log here for end without start
+		assert (false);
+	      }
+	    m_atomic_helper.unfix_atomic_replication_sequence (&thread_entry, header.trid);
+	    break;
+	  default:
+	    // do nothing
+	    break;
+	  }
+
+	{
+	  std::unique_lock<std::mutex> lock (m_redo_lsa_mutex);
+
+	  // better to be checked as soon as possible during the processing loop
+	  // however, this would need one more mutex lock; therefore, suffice to do it here
+	  assert (m_replication_active);
+
+	  m_redo_lsa = header.forw_lsa;
+	}
+
+	// to accurately track progress and avoid clients to wait for too long, notify each change
+	m_redo_lsa_condvar.notify_all ();
+
+	m_perfmon_redo_sync.track_and_start ();
+      }
+  }
+
+  template <typename T>
+  void
+  atomic_replicator::read_and_redo_record_with_atomic_consideration (cubthread::entry &thread_entry, LOG_RECTYPE rectype,
```

Comment:
We considered this approach and, although it could be an option, we favored inheritance. More differences will occur which might make the templated approach cluttered.